radio/src/targets/common/arm/stm32/flysky_gimbal_driver.cpp

```diff
@@ -0,0 +1,196 @@
+  0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
+  0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0
+};
+
+unsigned short calc_crc16(void *pBuffer,unsigned char BufferSize)
+{
+  unsigned short crc16;
+  crc16 = 0xffff;
+  while (BufferSize)
+  {
+    crc16 = (crc16 << 8) ^ CRC16Table[((crc16>>8) ^ (*(unsigned char *)pBuffer)) & 0x00ff];
+    pBuffer = (void *)((unsigned char *)pBuffer + 1);
+    BufferSize--;
+  }
+  return crc16;
+}
+
+uint16_t get_flysky_hall_adc_value(uint8_t ch)
+{
+  if (ch >= FLYSKY_HALL_CHANNEL_COUNT)
+  {
+    return 0;
+  }
+  return hall_adc_values[ch];
+}
+
+uint8_t HallGetByte(uint8_t * byte)
+{
+  return aux2SerialRxFifo.pop(*byte);
+}
+
+static void Parse_Character(STRUCT_HALL *hallBuffer, unsigned char ch)
+{
+  switch (hallBuffer->status) {
+    case GET_START:
+      if (FLYSKY_HALL_PROTOLO_HEAD == ch) {
+        hallBuffer->head = FLYSKY_HALL_PROTOLO_HEAD;
+        hallBuffer->status = GET_ID;
+        hallBuffer->msg_OK = 0;
+      }
+      break;
+    
+    case GET_ID:
+      hallBuffer->hallID.ID = ch;
+      hallBuffer->status = GET_LENGTH;
+      break;
+    
+    case GET_LENGTH:
+      hallBuffer->length = ch;
+      hallBuffer->dataIndex = 0;
+      hallBuffer->status = GET_DATA;
+      if (0 == hallBuffer->length) {
+        hallBuffer->status = GET_CHECKSUM;
+        hallBuffer->checkSum = 0;
+      }
+      break;
+    
+    case GET_DATA:
+      hallBuffer->data[hallBuffer->dataIndex++] = ch;
+      if (hallBuffer->dataIndex >= hallBuffer->length) {
+        hallBuffer->checkSum = 0;
+        hallBuffer->dataIndex = 0;
+        hallBuffer->status = GET_STATE;
+      }
+      break;
+    
+    case GET_STATE:
+      hallBuffer->checkSum = 0;
+      hallBuffer->dataIndex = 0;
+      hallBuffer->status = GET_CHECKSUM;
+    
+    case GET_CHECKSUM:
+      hallBuffer->checkSum |= ch << ((hallBuffer->dataIndex++) * 8);
+      if (hallBuffer->dataIndex >= 2) {
+        hallBuffer->dataIndex = 0;
+        hallBuffer->status = CHECKSUM;
+      } else {
+        break;
+      }
+    
+    case CHECKSUM:
+      if (hallBuffer->checkSum ==
+          calc_crc16((void *)&hallBuffer->head, hallBuffer->length + 3)) {
+        hallBuffer->msg_OK = 1;
+      }
+      hallBuffer->status = GET_START;
+      break; 
+  }
+}
+
+void flysky_gimbal_loop(void)
+{
+  uint8_t byte;
+
+  while (HallGetByte(&byte))
+  {
+    HallProtocol.index++;
+
+    Parse_Character(&HallProtocol, byte);
+    if (HallProtocol.msg_OK)
+    {
+      HallProtocol.msg_OK = 0;
+      HallProtocol.stickState = HallProtocol.data[HallProtocol.length - 1];
+
+      switch (HallProtocol.hallID.hall_Id.receiverID)
+      {
+        case TRANSFER_DIR_TXMCU:
+          if (HallProtocol.hallID.hall_Id.packetID == FLYSKY_HALL_RESP_TYPE_VALUES) {
+            int16_t* p_values = (int16_t*)HallProtocol.data;
+            for (uint8_t i = 0; i < 4; i++) {
+              hall_adc_values[i] = FLYSKY_OFFSET_VALUE - (p_values[i] >> 1);
```

Comment:
Please write directly to adcValues and remove changes in getADC(). I know it is causing conflict with existing gimballs sampling so in adcInit() remove sampling of standard gimballs related channels and writing to it when FLYSKY_GIMBAL is enabled.

Reply:
Yeah, I tried that but by disabling adc_channel_0, 1, 2, and 3, stuffs up the order of the switches, knobs and sticks. The best way would be to mask channels 0-3 like edgeTX does but could not find the similar adc code here

Reply:
Adjusting FIRST_ANALOG_ADC and NUM_ANALOGS_ADC does not help?

Reply:
No dice, even adjusting those it stuffs up the order of everything else.

Reply:
Ohh, actually, I havent changed the DMA_MemoryBaseAddr

Reply:
It is all now fixed